### PR TITLE
Build: No redundant npm installs on docker

### PR DESCRIPTION
**Pull request by sirreal**
I've noticed that our `install-if-deps-outdated` script often _does work_ when we build with Docker. This takes time and should be completely redundant after the `npm install --prod` step. If `npm-shrinkwrap.json` or `package.json` are modified, the docker image should be updated via an earlier layer. When the build step is reached, no installation should be necessary.

I inspected the `install-if-deps-outdated`, and it relies on the modification times of `npm-shrinkwrap.json` and `node_modules`.

https://github.com/Automattic/wp-calypso/blob/381850c162716063edc8785e5b45a58c796dc4ee/bin/install-if-deps-outdated.js#L16-L24

I believe this test is being thrown off by a subsequent `COPY`, overwriting the updated `npm-shrinkrarp` and affecting it's modification time.

https://github.com/Automattic/wp-calypso/blob/381850c162716063edc8785e5b45a58c796dc4ee/Dockerfile#L28-L38

This PR simply adds a `touch` to `npm-shrinkwrap.json` and `node_modules` to ensure that `install-if-deps-outdated` does not attempt install on build.

## Testing
* See https://github.com/Automattic/wp-calypso/pull/19796#issuecomment-345628115 for instructions on reliably reproducing the redundant install.
* Use `npm run build-docker`
* Confirm that time is spent in the `install-if-deps-outdated` step on `master`.
* Confirm that no time is spent in the `install-if-deps-outdated` step on this branch.
* Ensure that the produced image is correct (`npm run docker`).

### CODE REVIEW

**Comment by ockham** [comment]
Would be nice if `COPY` could exclude individual files :roll_eyes: Fix looks sane tho. Will give it a test (likely tomorrow).

**Comment by sirreal** [comment]
Also to note, the following `install` may be especially time consuming because `rm -rf /root/.npm` removes the local cache.

**Comment by sirreal** [comment]
> Would be nice if COPY could exclude individual files

I'm wondering if COPY should _explicitly include_ files. See https://github.com/Automattic/wp-calypso/pull/19801#issuecomment-344849818

**Comment by ockham** [comment]
> * Use `npm run build-docker`
> * Confirm that time is spent in the `install-if-deps-outdated` step on `master`.

I can't seem to confirm that on second (and subsequent) run(s). Even if I change part of the code for good measure, `npm run -s install-if-deps-outdated` doesn't do any work (only the actual build does).

**Comment by sirreal** [comment]
Thanks for testing @ockham 

> I can't seem to confirm that on second (and subsequent) run(s)

I'm having trouble reproducing now as well. On a recent build I observed this consistently but it's not clear what triggered it.

We can put this on hold, I'll be keeping an eye on this.

File: Dockerfile [hunk 1]
```diff
@@ -31,11 +31,27 @@ RUN        true \
...
+# `install-if-deps-outdated` should require no action.
+# However, time is being spent in the build step on
+# `install-if-deps-outdated`. This is because in the
+# following COPY, the npm-shrinkwrap mtime is being
+# updated, which is confusing `install-if-deps-outdated`.
+# Touch after copy to ensure that this layer will
+# not trigger additional install as part of the build
+# in the following step.
+COPY       . /calypso/
+RUN        touch npm-shrinkwrap.json node_modules
```
Comment (DanReyLop) [inline Dockerfile]: `touch`ing just `node_modules` would be better. `install-if-deps-updated` doesn't do any work if `node_modules` is **newer** than `shrinkwrap`.

Comment (sirreal) [reply Dockerfile]: 👍  Nice, thanks. 28881e4cf89228e4dd259d62730493317a80a475

**Comment by DanReyLop** [comment]
Even if it's not reproducible 100% of the time, I think this change should be merge. It doesn't hurt, and it helps in the cases where the bug was manifested.

I don't know if Docker's `COPY` preserves the file's `mtime`, maybe that could be making this harder to reproduce? If `COPY` doesn't preserve the `mtime`, then this should be *always* happening. Or almost always: If the second `COPY` manages to copy the `shrinkwrap` file less than a second after `npm install` ended, `install-if-deps-outdated` would **not** reinstall everything.

**Comment by sirreal** [comment]
@ockham I think I've found a way to reproduce reliably. We need to make sure there's a cached build for the installation step but not for the build step. You can try to abort the build when the build step is hit or you can modify the `Dockerfile` as follows:

```patch
diff --git a/Dockerfile b/Dockerfile
index fb92cb0ea6..548d443b4d 100644
--- a/Dockerfile
+++ b/Dockerfile
@@ -30,16 +30,3 @@ RUN        true \
            && npm install --production \
            && rm -rf /root/.npm \
            && true
-
-# Build the final layer
-#
-# This contains built environments of Calypso. It will
-# change any time any of the Calypso source-code changes.
-COPY       . /calypso/
-RUN        true \
-           && CALYPSO_ENV=production npm run build \
-           && chown -R nobody /calypso \
-           && true
-
-USER       nobody
-CMD        NODE_ENV=production node build/bundle.js
```

Now that you have a cached build for the npm install step, `touch npm-shrinkwrap.json` to ensure it's "newer" than the `node_modules` in the cached build. Now allow the build to continue through more steps and you should see that time is spent in `install-if-deps-outdated`.

I'm not sure how these conditions occur in reality, but I know I've experienced it. I imagine it's not difficult to encounter them when switching branches.

---

> I don't know if Docker's `COPY` preserves the file's `mtime`

@DanReyLop `mtime` should be maintained: 

> [An Individual file] is copied individually along with its metadata

https://docs.docker.com/engine/reference/builder/#copy

---

A note, this all relies on Docker's [build cache mechanism](https://docs.docker.com/engine/userguide/eng-image/dockerfile_best-practices/#build-cache). It compares instructions and checksums (ignoring `mtime`), which is why Docker knows it doesn't need to re-install, but our `install-if-deps-outdated` script erroneously re-installs.

**Comment by sirreal** [comment]
I've also observed this when switching branches and from a freshly cloned repo.

**Comment by sirreal** [comment]
I'm consistently seeing this which is adding significant time to my docker builds.

Could I get some feedback to get this in motion? 🙏

**Comment by sirreal** [comment]
It's worth noting that this difference can produce inconsistent builds. In one case, only production modules have been installed `npm install --production`. However, if `install-if-deps-outdated` does fire, then we've installed _all_ modules as the script runs `npm install`. Additionally, the "extra" modules increase build time for further steps, slowing down `chown` which is already the slowest step I've observed.

File: Dockerfile [hunk 2]
```diff
@@ -31,11 +31,27 @@ RUN        true \
            && rm -rf /root/.npm \
            && true
 
+# Build a "files" layer
```
Comment (samouri) [inline Dockerfile]: not a fan of "files".  maybe a "src" layer?

**Comment by samouri** [comment]
> Could I get some feedback to get this in motion?

this looks totally safe and the logic makes sense to me. I don't see why we wouldn't just merge this.

I'll do a couple tests locally to make sure build isn't busted

File: Dockerfile (hunk 2 above)
Comment (sirreal) [reply Dockerfile]: 👍  used "source" in d4096c3c2e6c2165c678acb2557eba67ab5f3ff9

**Pull request merged**